File: 0201 Bitwise AND of Numbers Range.java
class Solution {
    public int rangeBitwiseAnd(int m, int n) {
        int factor = 0;
        while (m != n){
            m >>= 1;
            n >>= 1;
            ++factor;
        }
        return m << factor;
    }
}

/* The other solution
class Solution {
    public int rangeBitwiseAnd(int m, int n) {
        int factor = 1;
        while (m != n){
            m >>= 1;
            n >>= 1;
            factor <<= 1;
        }
        return m * factor;
    }
}

class Solution {
    public int rangeBitwiseAnd(int m, int n) {
        while (n > m){
            n &= n-1;
        }
        return n;
    }
}
*/

File: 0201 Bitwise AND of Numbers Range.py
class Solution:
    def rangeBitwiseAnd(self, m: int, n: int) -> int:
        factor = 0
        while n != m:
            m >>= 1
            n >>= 1
            factor += 1
        return m << factor

''' other solutions
class Solution:
    def rangeBitwiseAnd(self, m: int, n: int) -> int:
        factor = 1
        while n != m:
            m >>= 1
            n >>= 1
            factor <<= 1
        return m * factor
        
class Solution:
    def rangeBitwiseAnd(self, m: int, n: int) -> int:
        while n > m:
            n &= n - 1
        return n        

'''